### ISSUE TITLE
[FEATURE REQUEST]: Enable requests of batches of experiments in Service API

### ISSUE THREAD

**Issue by CompRhys** [facebook/Ax#3120]
### Motivation

Currently the service API doesn't allow for batches of experiments to be requested unless they're sequential greedy strategies in which case it grabs experiments and feeds them into the `X_pending` state. There are some methods such as recently released qPO https://arxiv.org/abs/2410.06333 that are not compatible with this formulation. Otherwise (our current usecase) it would just be nice to be able to replicate the testing setup on the dev API where different arms are grouped into the same trial if they were requested at the same time. 

Our lab does batch processing of samples.

### Describe the solution you'd like to see implemented in Ax.

add an `n` argument to `AxClient.get_next_trial` or introduce an equivalent `get_next_trial_batch` that enables this functionality.

Happy to do this myself assuming that an interface/API acceptable to Ax maintainers is agreed.

### Describe any alternatives you've considered to the above solution.

Just using the service API's serial trials and assigning other metadata to indicate that these candidates formed part of the same batch.

### Code of Conduct

- [X] I agree to follow Ax's Code of Conduct

**Comment by Balandat** [comment]
This is something we generally want to support in our new APIs - @mpolson64 knows more about the plans and timelines.

**Comment by lena-kashtelyan** [comment]
Hi @CompRhys! Are you sure you need batches or would `AxClient.get_next_trials` do? What is the reason why you want to use `BatchTrial`-s specifically?

**Comment by CompRhys** [comment]
We're designing in batches and so given that it exists in the Ax schema in dev API it would be nice to match that in the service API. We can do it in service API by just leaving X_pending and then defining the batch idx in our own data schema.

So it's not needed but would be a nice to have.

**Comment by lena-kashtelyan** [comment]
In Ax, `BatchTrial`-s have a pretty specific meaning, you can read more about it here: https://ax.dev/docs/core#trial-vs-batch-trial. My sense is that your experiment doesn't match that definition, because the evaluation of each arm is in fact independent from one another, they are just done in parallel, is that right? 

If so, my suggestion would be to:
1. Use `get_next_trials` (not `get_next_trial`) to get a "batch" of trials all at once (we might need to correctly configure `max_parallelism` of the Ax' `GenerationStrategy` so it returns as many trials at once as you want –– let me know if you would like help with that),
2.  use the `metadata` field of `complete_trial` (https://github.com/facebook/Ax/blob/0f8660a10e42b84d0cc8c4b328c4d7ae51b5519c/ax/service/ax_client.py#L769) to track the index of the batch on your end. 

What do you think of this setup?

> We can do it in service API by just leaving X_pending

I'm not sure I understood this bit: why are you finding yourself needing to interact with `X_pending`? : ) `AxClient` should be handling the passing of all pending features, for you.

**Comment by CompRhys** [comment]
I can't remember the details as I haven't looked at what my coworker worked up in a while but I believe we we're handling X_pending as we're not keeping the process alive or using the Ax backend. When we trigger we were manually recreating the object and then needed to do something to reassign X_pending. If we were to call `get_next_trial` many times in a row in the same process then that would work as you say. I am not sure whether `get_next_trials` was already in the API when I posted this.

**Comment by lena-kashtelyan** [comment]
`get_next_trials` was in the API for a long time, but not featured in tutorials unfortunately, so it makes only the most sense that you didn't stumble into it! LMK if it doesn't solve the issue, I'll be glad to help more.

**Comment by VMLC-PV** [comment]
@CompRhys just curious, did you manage to implement qPO?

**Comment by sgbaird** [comment]
Maybe I'm interpreting incorrectly, but based on the logs it seems like `get_next_trials` is still doing fantasy modeling of some kind (discussion at https://github.com/facebook/Ax/issues/778#issuecomment-1013954552).

Reproducer: https://colab.research.google.com/drive/1zr2Xrom8dZrdBTiO2YAmbr76uAnlgZGi?usp=sharing

```python
[INFO 08-18 19:45:28] ax.service.ax_client: Starting optimization with verbose logging. To disable logging, set the `verbose_logging` argument to `False`. Note that float values in the logs are rounded to 6 decimal points.
[INFO 08-18 19:45:28] ax.service.utils.instantiation: Inferred value type of ParameterType.FLOAT for parameter x1. If that is not the expected value type, you can explicitly specify 'value_type' ('int', 'float', 'bool' or 'str') in parameter dict.
[INFO 08-18 19:45:28] ax.service.utils.instantiation: Inferred value type of ParameterType.FLOAT for parameter x2. If that is not the expected value type, you can explicitly specify 'value_type' ('int', 'float', 'bool' or 'str') in parameter dict.
[INFO 08-18 19:45:28] ax.service.utils.instantiation: Created search space: SearchSpace(parameters=[RangeParameter(name='x1', parameter_type=FLOAT, range=[-5.0, 10.0]), RangeParameter(name='x2', parameter_type=FLOAT, range=[0.0, 10.0])], parameter_constraints=[]).
[INFO 08-18 19:45:28] ax.modelbridge.dispatch_utils: Using Models.BOTORCH_MODULAR since there is at least one ordered parameter and there are no unordered categorical parameters.
[INFO 08-18 19:45:28] ax.modelbridge.dispatch_utils: Calculating the number of remaining initialization trials based on num_initialization_trials=None max_initialization_trials=None num_tunable_parameters=2 num_trials=None use_batch_trials=False
[INFO 08-18 19:45:28] ax.modelbridge.dispatch_utils: calculated num_initialization_trials=5
[INFO 08-18 19:45:28] ax.modelbridge.dispatch_utils: num_completed_initialization_trials=0 num_remaining_initialization_trials=5
[INFO 08-18 19:45:28] ax.modelbridge.dispatch_utils: `verbose`, `disable_progbar`, and `jit_compile` are not yet supported when using `choose_generation_strategy` with ModularBoTorchModel, dropping these arguments.
[INFO 08-18 19:45:28] ax.modelbridge.dispatch_utils: Using Bayesian Optimization generation strategy: GenerationStrategy(name='Sobol+BoTorch', steps=[Sobol for 5 trials, BoTorch for subsequent trials]). Iterations after 5 will take longer to generate due to model-fitting.
/usr/local/lib/python3.11/dist-packages/ax/modelbridge/cross_validation.py:439: UserWarning: Encountered exception in computing model fit quality: RandomModelBridge does not support prediction.
  warn("Encountered exception in computing model fit quality: " + str(e))
[INFO 08-18 19:45:28] ax.service.ax_client: Generated new trial 0 with parameters {'x1': 6.30547, 'x2': 7.604373} using model Sobol.
/usr/local/lib/python3.11/dist-packages/ax/modelbridge/cross_validation.py:439: UserWarning: Encountered exception in computing model fit quality: RandomModelBridge does not support prediction.
  warn("Encountered exception in computing model fit quality: " + str(e))
[INFO 08-18 19:45:28] ax.service.ax_client: Generated new trial 1 with parameters {'x1': -4.946741, 'x2': 4.201565} using model Sobol.
/usr/local/lib/python3.11/dist-packages/ax/modelbridge/cross_validation.py:439: UserWarning: Encountered exception in computing model fit quality: RandomModelBridge does not support prediction.
  warn("Encountered exception in computing model fit quality: " + str(e))
[INFO 08-18 19:45:28] ax.service.ax_client: Generated new trial 2 with parameters {'x1': 0.323003, 'x2': 5.739689} using model Sobol.
/usr/local/lib/python3.11/dist-packages/ax/modelbridge/cross_validation.py:439: UserWarning: Encountered exception in computing model fit quality: RandomModelBridge does not support prediction.
  warn("Encountered exception in computing model fit quality: " + str(e))
[INFO 08-18 19:45:28] ax.service.ax_client: Generated new trial 3 with parameters {'x1': 4.067933, 'x2': 2.335933} using model Sobol.
/usr/local/lib/python3.11/dist-packages/ax/modelbridge/cross_validation.py:439: UserWarning: Encountered exception in computing model fit quality: RandomModelBridge does not support prediction.
  warn("Encountered exception in computing model fit quality: " + str(e))
[INFO 08-18 19:45:28] ax.service.ax_client: Generated new trial 4 with parameters {'x1': 6.047932, 'x2': 6.64526} using model Sobol.
[INFO 08-18 19:45:28] ax.service.ax_client: Completed trial 0 with data: {'branin': (61.896536, None)}.
[INFO 08-18 19:45:28] ax.service.ax_client: Completed trial 1 with data: {'branin': (176.905239, None)}.
[INFO 08-18 19:45:28] ax.service.ax_client: Completed trial 2 with data: {'branin': (19.163292, None)}.
[INFO 08-18 19:45:28] ax.service.ax_client: Completed trial 3 with data: {'branin': (4.683646, None)}.
[INFO 08-18 19:45:28] ax.service.ax_client: Completed trial 4 with data: {'branin': (50.091287, None)}.
[INFO 08-18 19:45:29] ax.service.ax_client: Generated new trial 5 with parameters {'x1': 2.430566, 'x2': 2.768939} using model BoTorch.
/usr/local/lib/python3.11/dist-packages/ax/core/data.py:295: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.
  return cls(df=pd.concat(dfs, axis=0, sort=True))
[INFO 08-18 19:45:30] ax.service.ax_client: Generated new trial 6 with parameters {'x1': 2.497089, 'x2': 6.793068} using model BoTorch.
/usr/local/lib/python3.11/dist-packages/ax/core/data.py:295: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.
  return cls(df=pd.concat(dfs, axis=0, sort=True))
[INFO 08-18 19:45:32] ax.service.ax_client: Generated new trial 7 with parameters {'x1': 2.03841, 'x2': 0.0} using model BoTorch.
[INFO 08-18 19:45:32] ax.service.ax_client: Completed trial 5 with data: {'branin': (2.740398, None)}.
[INFO 08-18 19:45:32] ax.service.ax_client: Completed trial 6 with data: {'branin': (18.01983, None)}.
[INFO 08-18 19:45:32] ax.service.ax_client: Completed trial 7 with data: {'branin': (16.512643, None)}.
[INFO 08-18 19:45:33] ax.service.ax_client: Generated new trial 8 with parameters {'x1': 8.902525, 'x2': 0.0} using model BoTorch.
/usr/local/lib/python3.11/dist-packages/ax/core/data.py:295: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.
  return cls(df=pd.concat(dfs, axis=0, sort=True))
[INFO 08-18 19:45:34] ax.service.ax_client: Generated new trial 9 with parameters {'x1': 4.768437, 'x2': 0.0} using model BoTorch.
/usr/local/lib/python3.11/dist-packages/ax/core/data.py:295: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.
  return cls(df=pd.concat(dfs, axis=0, sort=True))
[INFO 08-18 19:45:35] ax.service.ax_client: Generated new trial 10 with parameters {'x1': 2.89575, 'x2': 2.919881} using model BoTorch.
[INFO 08-18 19:45:35] ax.service.ax_client: Completed trial 8 with data: {'branin': (5.961545, None)}.
[INFO 08-18 19:45:35] ax.service.ax_client: Completed trial 9 with data: {'branin': (12.355519, None)}.
[INFO 08-18 19:45:35] ax.service.ax_client: Completed trial 10 with data: {'branin': (0.884935, None)}.
```

To try to clarify a bit further, there is a delay between each generated trial that increases with the number of points. To me this would indicate fantasy modeling rather than an "all-at-once" joint acquisition approach.

https://github.com/user-attachments/assets/7bef750e-f55b-4aca-84ce-0869c34f25a0
